tests/base/hdf5_05.cc

```diff
@@ -0,0 +1,1463 @@
+    auto                       dataset =
+      group.create_dataset<Number>(dataset_name, dataset_dimensions);
+    dataset.set_query_io_mode(true);
+    deallog << "Dimensions " << dataset_name << "<" << type_name << ">"
+            << " (Write): " << dataset.get_dimensions() << std::endl;
+    deallog << "Size " << dataset_name << "<" << type_name << ">"
+            << " (Write): " << dataset.get_size() << std::endl;
+    deallog << "Rank " << dataset_name << "<" << type_name << ">"
+            << " (Write): " << dataset.get_rank() << std::endl;
+    auto data = initialize_container<std::vector, Number>(dataset_dimensions);
+    assign_data(data);
+    if (Utilities::MPI::this_mpi_process(mpi_communicator) == 0)
+      {
+        dataset.write(data);
+      }
+    else
+      {
+        dataset.template write_none<Number>();
+      }
+    deallog << "Data " + dataset_name << "<" << type_name << ">"
+            << " (Write): " << container_to_string(data) << std::endl;
+    pcout << "IO mode " + dataset_name << "<" << type_name << ">"
+          << " (Write): " << dataset.get_io_mode() << std::endl;
+    pcout << "Local no collective cause " + dataset_name << "<" << type_name
+          << ">"
+          << " (Write): " << dataset.get_local_no_collective_cause()
+          << std::endl;
+    pcout << "Global no collective cause " + dataset_name << "<" << type_name
+          << ">"
+          << " (Write): " << dataset.get_global_no_collective_cause()
+          << std::endl;
+  }
+
+  {
+    std::string dataset_name("dataset_3");
+
+    // This dataset is a 2x5x3 tensor
+    // Output generated by h5dump
+    //  (0,0,0): 0, 1, 2,         (1,0,0): 0, 2, 4,
+    //  (0,1,0): 3, 4, 5,         (1,1,0): 6, 8, 10,
+    //  (0,2,0): 6, 7, 8,         (1,2,0): 12, 14, 16,
+    //  (0,3,0): 9, 10, 11,       (1,3,0): 18, 20, 22,
+    //  (0,4,0): 12, 13, 14,      (1,4,0): 24, 26, 28
+
+    const std::vector<hsize_t> dataset_dimensions = {2, 5, 3};
+    auto                       dataset =
+      group.create_dataset<Number>(dataset_name, dataset_dimensions);
+    dataset.set_query_io_mode(true);
+    deallog << "Dimensions " << dataset_name << "<" << type_name << ">"
+            << " (Write): " << dataset.get_dimensions() << std::endl;
+    deallog << "Size " << dataset_name << "<" << type_name << ">"
+            << " (Write): " << dataset.get_size() << std::endl;
+    deallog << "Rank " << dataset_name << "<" << type_name << ">"
+            << " (Write): " << dataset.get_rank() << std::endl;
+
+    if (Utilities::MPI::this_mpi_process(mpi_communicator) == 0)
+      {
+        const std::vector<hsize_t> data_dimensions = {5, 3};
+        auto data = initialize_container<std::vector, Number>(data_dimensions);
+        assign_data(data);
+        const std::vector<hsize_t> hyperslab_offset = {0, 0, 0};
+        const std::vector<hsize_t> hyperslab_count  = {1, 5, 3};
+        dataset.write_hyperslab(data, hyperslab_offset, hyperslab_count);
+
```

Comment:
Here I write a a hyperslab (sub-matrix) with rank 2 in a dataset of rank 3. This can be seen as  a slice (rank 2) of a cube (rank 3). For example each slice correspond to a different time

Note that `std::vector` and Vector can be used to write hypersalbs of rank 1,2,3,4,...